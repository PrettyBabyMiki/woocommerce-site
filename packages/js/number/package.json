{
	"name": "@woocommerce/number",
	"version": "2.3.0",
	"description": "Number formatting utilities for WooCommerce.",
	"author": "Automattic",
	"license": "GPL-3.0-or-later",
	"keywords": [
		"wordpress",
		"woocommerce"
	],
	"homepage": "https://github.com/woocommerce/woocommerce/tree/trunk/packages/js/number/README.md",
	"repository": {
		"type": "git",
		"url": "https://github.com/woocommerce/woocommerce.git"
	},
	"bugs": {
		"url": "https://github.com/woocommerce/woocommerce/issues"
	},
	"main": "build/index.js",
	"module": "build-module/index.js",
	"types": "build-types",
	"react-native": "src/index",
	"dependencies": {
		"locutus": "^2.0.16"
	},
	"publishConfig": {
		"access": "public"
	},
	"scripts": {
		"postinstall": "composer install",
		"changelog": "composer exec -- changelogger",
		"clean": "pnpm exec rimraf tsconfig.tsbuildinfo build build-*",
		"build": "tsc --build ./tsconfig.json ./tsconfig-cjs.json",
		"start": "tsc --build --watch",
		"prepack": "pnpm run clean && pnpm run build",
		"lint": "eslint src",
		"lint:fix": "eslint src --fix",
		"test": "jest --config ./jest.config.json",
		"test-staged": "jest --bail --config ./jest.config.json --findRelatedTests"
	},
	"devDependencies": {
		"@babel/core": "^7.17.5",
		"@woocommerce/eslint-plugin": "workspace:*",
		"eslint": "^8.12.0",
		"@babel/runtime": "^7.17.2",
		"jest": "^27.5.1",
		"jest-cli": "^27.5.1",
		"rimraf": "^3.0.2",
		"ts-jest": "^27.1.3",
		"typescript": "^4.6.2"
	},
	"lint-staged": {
		"*.(t|j)s?(x)": [
			"pnpm lint:fix",
			"pnpm test-staged"
		]
	}
}
