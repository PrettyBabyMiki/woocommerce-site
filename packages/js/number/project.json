{
     "root": "packages/js/number",
     "sourceRoot": "packages/js/number/src",
     "projectType": "library",
     "targets": {
         "changelog": {
             "executor": "./tools/executors/changelogger:changelog",
             "options": {
                 "action": "add",
                 "cwd": "packages/js/number"
             }
         },
         "build": {
             "executor": "@nrwl/workspace:run-script",
             "options": {
                 "script": "build"
             }
         },
         "build-watch": {
			"executor": "@nrwl/workspace:run-script",
			"options": {
                "script": "start"
			}
		},
         "test": {
             "executor": "@nrwl/workspace:run-script",
             "options": {
                 "script": "test"
             }
         }
     }
 }