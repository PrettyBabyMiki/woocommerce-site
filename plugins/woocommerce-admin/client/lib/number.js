/**
 * Formats a number using site's current locale
 *
 * @format
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat
 * @param {Number|String}  number    number to format
 * @returns {?String}                  A formatted string.
 */

export function numberFormat( number ) {
	const locale = wcSettings.siteLocale || 'en-US'; // Default so we don't break.

	if ( 'number' !== typeof number ) {
		number = parseFloat( number );
	}
	if ( isNaN( number ) ) {
		return '';
	}
	return new Intl.NumberFormat( locale ).format( number );
}
