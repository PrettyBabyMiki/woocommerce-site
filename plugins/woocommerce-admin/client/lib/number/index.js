/** @format */
/**
 * External dependencies
 */
import { get } from 'lodash';
const number_format = require( 'locutus/php/strings/number_format' );

/**
 * Formats a number using site's current locale
 *
 * @see http://locutus.io/php/strings/number_format/
 * @param {Number|String} number number to format
 * @param {int|null} [precision=null] optional decimal precision
 * @returns {?String} A formatted string.
 */

export function numberFormat( number, precision = null ) {
	if ( 'number' !== typeof number ) {
		number = parseFloat( number );
	}

	if ( isNaN( number ) ) {
		return '';
	}

	const decimalSeparator = get( wcSettings, [ 'currency', 'decimal_separator' ], '.' );
	const thousandSeparator = get( wcSettings, [ 'currency', 'thousand_separator' ], ',' );
	precision = parseInt( precision );

	if ( isNaN( precision ) ) {
		const [ , decimals ] = number.toString().split( '.' );
		precision = decimals ? decimals.length : 0;
	}

	return number_format( number, precision, decimalSeparator, thousandSeparator );
}
