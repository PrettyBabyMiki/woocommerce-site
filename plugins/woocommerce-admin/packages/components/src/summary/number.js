/**
 * External dependencies
 */
import {
	Button,
	Tooltip,
	__experimentalText as Text,
} from '@wordpress/components';
import { sprintf, __ } from '@wordpress/i18n';
import classnames from 'classnames';
import Gridicon from 'gridicons';
import { isNil, noop } from 'lodash';
import PropTypes from 'prop-types';
/**
 * Internal dependencies
 */
import Link from '../link';

/**
 * A component to show a value, label, and an optional change percentage. Can also act as a link to a specific report focus.
 *
 * @return {Object} -
 */
const SummaryNumber = ( {
	delta,
	href,
	hrefType,
	isOpen,
	label,
	onToggle,
	prevLabel,
	prevValue,
	reverseTrend,
	selected,
	value,
	onLinkClickCallback,
	isHomescreen,
} ) => {
	const liClasses = classnames( 'woocommerce-summary__item-container', {
		'is-homescreen': isHomescreen,
		'is-dropdown-button': onToggle,
		'is-dropdown-expanded': isOpen,
	} );
	const classes = classnames( 'woocommerce-summary__item', {
		'is-selected': selected,
		'is-good-trend': reverseTrend ? delta < 0 : delta > 0,
		'is-bad-trend': reverseTrend ? delta > 0 : delta < 0,
	} );

	let screenReaderLabel =
		delta > 0
			? sprintf(
					__( 'Up %d%% from %s', 'woocommerce-admin' ),
					delta,
					prevLabel
			  )
			: sprintf(
					__( 'Down %d%% from %s', 'woocommerce-admin' ),
					Math.abs( delta ),
					prevLabel
			  );
	if ( ! delta ) {
		screenReaderLabel = sprintf(
			__( 'No change from %s', 'woocommerce-admin' ),
			prevLabel
		);
	}

	let Container;
	const containerProps = {
		className: classes,
		'aria-current': selected ? 'page' : null,
	};

	if ( onToggle || href ) {
		const isButton = !! onToggle;
		Container = isButton ? Button : Link;
		if ( isButton ) {
			containerProps.onClick = onToggle;
			containerProps[ 'aria-expanded' ] = isOpen;
		} else {
			containerProps.href = href;
			containerProps.role = 'menuitem';
			containerProps.onClick = onLinkClickCallback;
			containerProps.type = hrefType;
		}
	} else {
		Container = 'div';
	}

	return (
		<li className={ liClasses }>
			<Container { ...containerProps }>
				<div className="woocommerce-summary__item-label">
					<Text variant="body.small">{ label }</Text>
				</div>

				<div className="woocommerce-summary__item-data">
					<div className="woocommerce-summary__item-value">
						<Text variant="title.small">
							{ ! isNil( value )
								? value
								: __( 'N/A', 'woocommerce-admin' ) }
						</Text>
					</div>

					<Tooltip
						text={
							! isNil( prevValue )
								? `${ prevLabel } ${ prevValue }`
								: __( 'N/A', 'woocommerce-admin' )
						}
						position="top center"
					>
						<div
							className="woocommerce-summary__item-delta"
							role="presentation"
							aria-label={ screenReaderLabel }
						>
							<Text variant="caption">
								{ ! isNil( delta )
									? sprintf(
											__( '%d%%', 'woocommerce-admin' ),
											delta
									  )
									: __( 'N/A', 'woocommerce-admin' ) }
							</Text>
						</div>
					</Tooltip>
				</div>
				{ onToggle ? (
					<Gridicon
						className="woocommerce-summary__toggle"
						icon="chevron-down"
						size={ 24 }
					/>
				) : null }
			</Container>
		</li>
	);
};

SummaryNumber.propTypes = {
	/**
	 * A number to represent the percentage change since the last comparison period - positive numbers will show
	 * a green up arrow, negative numbers will show a red down arrow, and zero will show a flat right arrow.
	 * If omitted, no change value will display.
	 */
	delta: PropTypes.number,
	/**
	 * An internal link to the report focused on this number.
	 */
	href: PropTypes.string,
	/**
	 * The type of the link
	 */
	hrefType: PropTypes.oneOf( [ 'wp-admin', 'wc-admin', 'external' ] )
		.isRequired,
	/**
	 * Boolean describing whether the menu list is open. Only applies in mobile view,
	 * and only applies to the toggle-able item (first in the list).
	 */
	isOpen: PropTypes.bool,
	/**
	 * A string description of this value, ex "Revenue", or "New Customers"
	 */
	label: PropTypes.string.isRequired,
	/**
	 * A function used to switch the given SummaryNumber to a button, and called on click.
	 */
	onToggle: PropTypes.func,
	/**
	 * A string description of the previous value's timeframe, ex "Previous Year:".
	 */
	prevLabel: PropTypes.string,
	/**
	 * A string or number value to display - a string is allowed so we can accept currency formatting.
	 * If omitted, this section won't display.
	 */
	prevValue: PropTypes.oneOfType( [ PropTypes.number, PropTypes.string ] ),
	/**
	 * A boolean used to indicate that a negative delta is "good", and should be styled like a positive (and vice-versa).
	 */
	reverseTrend: PropTypes.bool,
	/**
	 * A boolean used to show a highlight style on this number.
	 */
	selected: PropTypes.bool,
	/**
	 * A string or number value to display - a string is allowed so we can accept currency formatting.
	 */
	value: PropTypes.oneOfType( [ PropTypes.number, PropTypes.string ] ),
	/**
	 * A function to be called after a SummaryNumber, rendered as a link, is clicked.
	 */
	onLinkClickCallback: PropTypes.func,
};

SummaryNumber.defaultProps = {
	href: '',
	hrefType: 'wc-admin',
	isOpen: false,
	prevLabel: __( 'Previous Period:', 'woocommerce-admin' ),
	reverseTrend: false,
	selected: false,
	onLinkClickCallback: noop,
};

export default SummaryNumber;
