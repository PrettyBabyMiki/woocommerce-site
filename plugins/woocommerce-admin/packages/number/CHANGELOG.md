# 1.0.3

- Update liscence to GPL-3.0-or-later

# 1.0.2

- Bump dependency versions.

# 1.0.1

# 1.0.0

- Initial release exports `numberFormat`, `formatValue`, and `calculateDelta`
