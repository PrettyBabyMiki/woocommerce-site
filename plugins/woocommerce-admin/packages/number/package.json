{
	"name": "@woocommerce/number",
	"version": "1.0.2",
	"description": "Number formatting utilities for WooCommerce.",
	"author": "Automattic",
	"license": "GPL-2.0-or-later",
	"keywords": [
		"wordpress",
		"woocommerce"
	],
	"homepage": "https://github.com/woocommerce/woocommerce-admin/tree/master/packages/number/README.md",
	"repository": {
		"type": "git",
		"url": "https://github.com/woocommerce/woocommerce-admin.git"
	},
	"bugs": {
		"url": "https://github.com/woocommerce/woocommerce-admin/issues"
	},
	"main": "build/index.js",
	"module": "build-module/index.js",
	"react-native": "src/index",
	"dependencies": {
		"@babel/runtime-corejs2": "7.4.4",
		"locutus": "2.0.10",
		"lodash": "4.17.11"
	},
	"publishConfig": {
		"access": "public"
	}
}
