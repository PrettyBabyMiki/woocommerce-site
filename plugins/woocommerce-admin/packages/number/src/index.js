/** @format */
/* eslint-disable wpcalypso/import-docblock */

const number_format = require( 'locutus/php/strings/number_format' );

/**
 * Formats a number using site's current locale
 *
 * @see http://locutus.io/php/strings/number_format/
 * @param {Object} numberConfig number formatting configuration object.
 * @param {Number|String} number number to format
 * @returns {?String} A formatted string.
 */
export function numberFormat( {
	precision = null,
	decimalSeparator = '.',
	thousandSeparator = ',',
}, number ) {
	if ( 'number' !== typeof number ) {
		number = parseFloat( number );
	}

	if ( isNaN( number ) ) {
		return '';
	}

	precision = parseInt( precision );

	if ( isNaN( precision ) ) {
		const [ , decimals ] = number.toString().split( '.' );
		precision = decimals ? decimals.length : 0;
	}

	return number_format( number, precision, decimalSeparator, thousandSeparator );
}

/**
 * Formats a number string based on type of `average` or `number`.
 *
 * @param {Object} numberConfig number formatting configuration object.
 * @param {String} type of number to format, average or number
 * @param {int} value to format.
 * @returns {?String} A formatted string.
 */
export function formatValue( numberConfig, type, value ) {
	if ( ! Number.isFinite( value ) ) {
		return null;
	}

	switch ( type ) {
		case 'average':
			return Math.round( value );
		case 'number':
			return numberFormat( { ...numberConfig, precision: null }, value );
	}
}

/**
 * Calculates the delta/percentage change between two numbers.
 *
 * @param {int} primaryValue the value to calculate change for.
 * @param {int} secondaryValue the baseline which to calculdate the change against.
 * @returns {?int} Percent change between the primaryValue from the secondaryValue.
 */
export function calculateDelta( primaryValue, secondaryValue ) {
	if ( ! Number.isFinite( primaryValue ) || ! Number.isFinite( secondaryValue ) ) {
		return null;
	}

	if ( secondaryValue === 0 ) {
		return 0;
	}

	return Math.round( ( primaryValue - secondaryValue ) / secondaryValue * 100 );
}
