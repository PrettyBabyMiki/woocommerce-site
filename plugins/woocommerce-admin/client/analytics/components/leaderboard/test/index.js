/**
 * External dependencies
 */
import { mount, shallow } from 'enzyme';
import { numberFormat } from '@woocommerce/number';
import CurrencyFactory from '@woocommerce/currency';
import { CURRENCY } from '@woocommerce/wc-admin-settings';

/**
 * Internal dependencies
 */
import { Leaderboard } from '../';
import mockData from '../data/top-selling-products-mock-data';

const { formatAmount, formatDecimal } = CurrencyFactory( CURRENCY );

const rows = mockData.map( ( row ) => {
	const {
		name,
		items_sold: itemsSold,
		net_revenue: netRevenue,
		orders_count: ordersCount,
	} = row;
	return [
		{
			display: '<a href="#">' + name + '</a>',
			value: name,
		},
		{
			display: numberFormat( CURRENCY, itemsSold ),
			value: itemsSold,
		},
		{
			display: numberFormat( CURRENCY, ordersCount ),
			value: ordersCount,
		},
		{
			display: formatAmount( netRevenue ),
			value: formatDecimal( netRevenue ),
		},
	];
} );

const headers = [
	{
		label: 'Name',
	},
	{
		label: 'Items Sold',
	},
	{
		label: 'Orders',
	},
	{
		label: 'Net Sales',
	},
];

describe( 'Leaderboard', () => {
	test( 'should render empty message when there are no rows', () => {
		const leaderboard = shallow(
			<Leaderboard
				id="products"
				title={ '' }
				headers={ [] }
				rows={ [] }
				totalRows={ 5 }
			/>
		);

		expect( leaderboard.find( 'EmptyTable' ).length ).toBe( 1 );
	} );

	test( 'should render correct data in the table', () => {
		const leaderboard = mount(
			<Leaderboard
				id="products"
				title={ '' }
				headers={ headers }
				rows={ rows }
				totalRows={ 5 }
			/>
		);
		const table = leaderboard.find( 'TableCard' );
		const firstRow = table.props().rows[ 0 ];
		const tableItems = leaderboard.find( '.woocommerce-table__item' );

		expect( firstRow[ 0 ].value ).toBe( mockData[ 0 ].name );
		expect( firstRow[ 1 ].value ).toBe( mockData[ 0 ].items_sold );
		expect( firstRow[ 2 ].value ).toBe( mockData[ 0 ].orders_count );
		expect( firstRow[ 3 ].value ).toBe(
			formatDecimal( mockData[ 0 ].net_revenue )
		);

		expect(
			leaderboard.render().find( '.woocommerce-table__item a' ).length
		).toBe( 5 );
		expect( tableItems.at( 0 ).text() ).toBe( mockData[ 0 ].name );
		expect( tableItems.at( 1 ).text() ).toBe(
			numberFormat( CURRENCY, mockData[ 0 ].items_sold )
		);
		expect( tableItems.at( 2 ).text() ).toBe(
			numberFormat( CURRENCY, mockData[ 0 ].orders_count )
		);
		expect( tableItems.at( 3 ).text() ).toBe(
			formatAmount( mockData[ 0 ].net_revenue )
		);
	} );
} );
