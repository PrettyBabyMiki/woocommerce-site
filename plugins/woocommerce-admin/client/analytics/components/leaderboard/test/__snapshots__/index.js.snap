// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard should render correct data in the table 1`] = `
<div>
  <div
    class="woocommerce-card woocommerce-table woocommerce-analytics__card woocommerce-leaderboard"
  >
    
    <div
      class="woocommerce-card__body"
    >
      <div
        aria-hidden="false"
        aria-labelledby="caption-0"
        class="woocommerce-table__table"
        role="group"
      >
        <table>
          <caption
            class="woocommerce-table__caption screen-reader-text"
            id="caption-0"
          >
            
          </caption>
          <tbody>
            <tr>
              <th
                class="woocommerce-table__header is-left-aligned"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Name
                </span>
              </th>
              <th
                class="woocommerce-table__header"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Items Sold
                </span>
              </th>
              <th
                class="woocommerce-table__header"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Orders
                </span>
              </th>
              <th
                class="woocommerce-table__header"
                role="columnheader"
                scope="col"
              >
                <span
                  aria-hidden="false"
                >
                  Net Sales
                </span>
              </th>
            </tr>
            <tr>
              <th
                class="woocommerce-table__item is-left-aligned"
                scope="row"
              >
                <div>
                  <a
                    href="#"
                  >
                    awesome shirt
                  </a>
                </div>
              </th>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  1,000.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  54.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  $999.99
                </div>
              </td>
            </tr>
            <tr>
              <th
                class="woocommerce-table__item is-left-aligned"
                scope="row"
              >
                <div>
                  <a
                    href="#"
                  >
                    awesome pants
                  </a>
                </div>
              </th>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  90.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  41.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  $875.00
                </div>
              </td>
            </tr>
            <tr>
              <th
                class="woocommerce-table__item is-left-aligned"
                scope="row"
              >
                <div>
                  <a
                    href="#"
                  >
                    awesome hat
                  </a>
                </div>
              </th>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  55.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  28.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  $75.75
                </div>
              </td>
            </tr>
            <tr>
              <th
                class="woocommerce-table__item is-left-aligned"
                scope="row"
              >
                <div>
                  <a
                    href="#"
                  >
                    awesome sticker
                  </a>
                </div>
              </th>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  10.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  14.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  $24.50
                </div>
              </td>
            </tr>
            <tr>
              <th
                class="woocommerce-table__item is-left-aligned"
                scope="row"
              >
                <div>
                  <a
                    href="#"
                  >
                    awesome button
                  </a>
                </div>
              </th>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  1.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  1.00
                </div>
              </td>
              <td
                class="woocommerce-table__item"
              >
                <div>
                  $0.99
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Leaderboard should render empty message when there are no rows 1`] = `
<div>
  <div
    class="woocommerce-card woocommerce-leaderboard"
  >
    
    <div
      class="woocommerce-card__body"
    >
      <div
        class="woocommerce-table is-empty"
        style="--number-of-rows: 5;"
      >
        No data recorded for the selected time period.
      </div>
    </div>
  </div>
</div>
`;
